lib/ecto/changeset.ex

```diff
@@ -1786,42 +1786,43 @@ defmodule Ecto.Changeset do
   """
   def unsafe_validate_unique(changeset, fields, repo, opts \\ []) when is_list(opts) do
     fields = List.wrap(fields)
-    {validations, struct} =
+
+    {validations, schema} =
       case changeset do
-        %Ecto.Changeset{validations: validations, data: %{__struct__: struct}} ->
-          {validations, struct}
+        %Ecto.Changeset{validations: validations, data: %schema{}} ->
+          {validations, schema}
         %Ecto.Changeset{} ->
           raise ArgumentError, "unsafe_validate_unique/4 does not work with schemaless changesets"
       end
-    changeset = %{changeset | validations: [{:unsafe_unique, fields} | validations]}
 
-    where_clause = for field <- fields do
-      {field, get_field(changeset, field)}
-    end
+    changeset = %{changeset | validations: [{:unsafe_unique, fields} | validations]}
 
-    # No need to query if we haven't changed any of the fields in question
-    unrelated_changes? = fields -- Map.keys(changeset.changes) == fields
+    {skip_validation?, where_clause} =
+      Enum.reduce_while(fields, {true, []}, fn field, {skip_validation?, where_clause} ->
+        case fetch_field(changeset, field) do
+          {tag, value} ->
+            if is_nil(value) do
+              # If any field is not present, we cannot query for uniqueness.
+              {:halt, {true, []}}
+            else
+              # If no field has changed, there is no need to validate.
+              {:cont,
+               {skip_validation? and tag != :changes,
+                [{field, value} | where_clause]}}
+            end
 
-    # If we don't have values for all fields, we can't query for uniqueness
-    any_nil_values_for_fields? = Enum.any?(where_clause, &(&1 |> elem(1) |> is_nil()))
+          :error ->
+            # If any field is not present, we cannot query for uniqueness.
+            {:halt, {true, []}}
+        end
+      end)
 
-    if unrelated_changes? || any_nil_values_for_fields? do
+    if skip_validation? do
```

Comment:
I find the previous approach much more readable, so I would propose to keep this part the same. Perhaps simply rewrite:

```
unrelated_changes? = fields -- Map.keys(changeset.changes) == fields
```

to 

```
unrelated_changes? = Enum.all?(fields, &not Map.has_key?(changeset.changes, &1))
```

I am not super worried about performance because fields is going to be 5 elements at worse, so the multiple traversals do not really matter. :)

Reply:
yeah, I don't feel very strong about that too.